optimizely/event/chan_queue.go

```diff
@@ -0,0 +1,68 @@
+package event
+
+var done = make(chan bool)
+
+type ChanQueue struct {
+	ch chan UserEvent
+	messages Queue
+}
+
+// Get returns queue for given count size
+func (i *ChanQueue) Get(count int) []interface{} {
+
+	events := []interface{}{}
+
+	messages := i.messages.Get(count)
+	for _, message := range messages {
+		events = append(events, message)
+	}
+
+	return events
+}
+
+func (i *ChanQueue) publish(userEvent UserEvent) {
+	i.ch <- userEvent
+}
+
+// Add appends item to queue
+func (i *ChanQueue) Add(item interface{}) {
+	event, ok := item.(UserEvent)
+	if !ok {
+		// cannot add non-user events
+		return
+	}
+
+	i.publish(event)
+}
+
+// Remove removes item from queue and returns elements slice
+func (i *ChanQueue) Remove(count int) []interface{} {
+	userEvents := make([]interface{},0, count)
+	events := i.messages.Remove(count)
+	for _,message := range events {
+		userEvents = append(userEvents, message)
+	}
+	return userEvents
+}
+
+// Size returns size of queue
+func (i *ChanQueue) Size() int {
+	return i.messages.Size()
+}
+
+// NewNSQueue returns new NSQ based queue with given queueSize
+func NewChanQueue(queueSize int) Queue {
+
+	ch := make(chan UserEvent)
+
+	i := &ChanQueue{ch:ch, messages: NewInMemoryQueue(queueSize)}
+
+	go func() {
+		for item := range i.ch {
```

Comment:
Any thoughts on this @pawels-optimizely and @thomaszurkan-optimizely ?

Reply:
ah, yes, good point, unless it will be invoked from something that has a context already (like polling manager) but I doubt it.  do we still have different implementations of the queue?

Reply:
and it that case, for loop will need to be rewritten as select, like i did in polling manager ( it will be same logic)

Reply:
I pulled out the context checking into the event processor and dispatcher instead of the queue.  I'm wondering if the queue interface needs a close or on close method.

Reply:
Ok I think that should be ok then.